In [1]:
!pip install sklearn


  Running setup.py bdist_wheel for sklearn ... - done
  Stored in directory: /root/.cache/pip/wheels/76/03/bb/589d421d27431bcd2c6da284d5f2286c8e3b2ea3cf1594c074
Successfully built sklearn


In [0]:
import pandas as pd  
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score, roc_auc_score

In [0]:
paths = 'http://bigdata.cs.byu.edu/Lectures/EEGEyeState.csv'
df = pd.read_table(paths,sep=',',encoding='latin1')

In [0]:
y = df['eyeDetection']
X = df.iloc[:,:-1]

In [5]:
print(len(y))
print(len(X.O1))

14980
14980


In [0]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=.3,random_state=40)

In [0]:
scaler = StandardScaler()

In [8]:
scaler.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [0]:
X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test)

In [0]:
mlp = MLPClassifier(activation='relu',hidden_layer_sizes=(12,12,12),max_iter=500)

In [11]:
mlp.fit(X_train,y_train)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(12, 12, 12), learning_rate='constant',
       learning_rate_init=0.001, max_iter=500, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [0]:
nn_preds = mlp.predict(X_test)

In [13]:
conf_mat = pd.crosstab(y_test,nn_preds,margins=True)
conf_mat

col_0            0    1   All
eyeDetection                 
0             2063  413  2476
1             1787  231  2018
All           3850  644  4494

In [14]:
#sensitivity 
print('Our Sensitivity value is:', conf_mat.iloc[1,1]/conf_mat.iloc[1,2])
#specificity
print('Our Specificity is:',conf_mat.iloc[0,0]/conf_mat.iloc[0,2])
#positive predictive value 
print('Our PPV is:',conf_mat.iloc[1,1]/conf_mat.iloc[2,1])
#negative predictive value 
print('Our NPV is:',conf_mat.iloc[0,0]/conf_mat.iloc[2,0])
#accuracy
print('Our accuracy score is:',accuracy_score(y_test,nn_preds))

Our Sensitivity value is: 0.11446977205153618
Our Specificity is: 0.8331987075928917
Our PPV is: 0.358695652173913
Our NPV is: 0.5358441558441558
Our accuracy score is: 0.5104583889630618


In [15]:

roc_auc_score(y_test, nn_preds)

0.47383423982221395

# Summary

Our neural network results indicate an accuracy over 55%. This suggests fewer than 45% accuracy when predicting values out of our trained data. 

From the confusion matrix we observe specificity near 76%, indicating the ability to predict eye state when it is not open. Sensitivity, on the other hand, was found to be near 30% suggesting the model’s capacity to predict whether the eye was open when it truly was. As the aim is to maximize sensitivity, 30% may not be the optimum model for this task.

Another metric that is used to measure the performance of this model is the area under the curve: 56%. Though this is slightly better than a coin flip, this particular neural network model does not perform as well as other models for this particular analysis. This may be due to the simple composition of the network with 3 layers and 12 hidden nodes. For this model to determine utility, a more complex model (ie. RNN) may need to be considered.

